pkg/volume/azure_dd/azure_dd.go

```diff
@@ -0,0 +1,348 @@
+	// azures used directly in a pod have a ReadOnly flag set by the pod author.
+	// azures used as a PersistentVolume gets the ReadOnly flag indirectly through the persistent-claim volume used to mount the PV
+	azure, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	fsType := *azure.FSType
+	diskName := azure.DiskName
+	diskUri := azure.DataDiskURI
+	cachingMode := *azure.CachingMode
+	return &azureDiskMounter{
+		azureDisk: &azureDisk{
+			podUID:      podUID,
+			volName:     spec.Name(),
+			diskName:    diskName,
+			diskUri:     diskUri,
+			cachingMode: cachingMode,
+			namespace:   namespace,
+			mounter:     mounter,
+			plugin:      plugin,
+		},
+		fsType:      fsType,
+		readOnly:    *azure.ReadOnly,
+		diskMounter: &mount.SafeFormatAndMount{Interface: plugin.host.GetMounter(), Runner: exec.New()}}, nil
+}
+
+func (plugin *azureDataDiskPlugin) NewUnmounter(volName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(volName, podUID, plugin.host.GetMounter())
+}
+
+func (plugin *azureDataDiskPlugin) newUnmounterInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Unmounter, error) {
+	return &azureDiskUnmounter{
+		&azureDisk{
+			podUID:  podUID,
+			volName: volName,
+			mounter: mounter,
+			plugin:  plugin,
+		},
+	}, nil
+}
+
+type azureDisk struct {
+	volName     string
+	podUID      types.UID
+	diskName    string
+	diskUri     string
+	cachingMode api.AzureDataDiskCachingMode
+	namespace   string
+	mounter     mount.Interface
+	plugin      *azureDataDiskPlugin
+	volume.MetricsNil
+}
+
+type azureDiskMounter struct {
+	*azureDisk
+	// Filesystem type, optional.
+	fsType string
+	// Specifies whether the disk will be attached as read-only.
+	readOnly bool
+	// diskMounter provides the interface that is used to mount the actual block device.
+	diskMounter *mount.SafeFormatAndMount
+}
+
+var _ volume.Mounter = &azureDiskMounter{}
+
+func (b *azureDiskMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        b.readOnly,
+		Managed:         !b.readOnly,
+		SupportsSELinux: true,
+	}
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *azureDiskMounter) SetUp(fsGroup *int64) error {
+	return b.SetUpAt(b.GetPath(), fsGroup)
+}
+
+// SetUpAt attaches the disk and bind mounts to the volume path.
+func (b *azureDiskMounter) SetUpAt(dir string, fsGroup *int64) error {
+	b.plugin.volumeLocks.LockKey(b.diskName)
+	defer b.plugin.volumeLocks.UnlockKey(b.diskName)
+
+	// TODO: handle failed mounts here.
+	notMnt, err := b.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("DataDisk set up: %s %v %v", dir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("IsLikelyNotMountPoint failed: %v", err)
+		return err
+	}
+	if !notMnt {
+		glog.V(4).Infof("%s is a mount point", dir)
+		return nil
```

Comment:
the contract is that if it is a mountpoint, then the volume is already mounted, and no more work to do.